p11-kit/rpc-client.c

```diff
@@ -1883,6 +1884,379 @@ rpc_C_GenerateRandom (CK_X_FUNCTION_LIST *self,
+		IN_BYTE_ARRAY (parameter, parameter_len)
+		IN_BYTE_ARRAY (ciphertext_part, ciphertext_part_len)
+		IN_BYTE_BUFFER (plaintext_part, plaintext_part_len);
+		IN_ULONG (flags)
+	PROCESS_CALL;
+		OUT_BYTE_ARRAY (plaintext_part, plaintext_part_len)
+	END_CALL;
+}
+
+static CK_RV
+rpc_C_MessageDecryptFinal (CK_X_FUNCTION_LIST *self,
+                           CK_SESSION_HANDLE session)
+{
+	BEGIN_CALL_OR (C_MessageDecryptFinal, self, CKR_SESSION_HANDLE_INVALID);
+		IN_ULONG (session)
+	PROCESS_CALL;
+	END_CALL;
+}
+
+static CK_RV
+rpc_C_MessageSignInit (CK_X_FUNCTION_LIST *self,
+                       CK_SESSION_HANDLE session,
+                       CK_MECHANISM_PTR mechanism,
+                       CK_OBJECT_HANDLE key)
+{
+	BEGIN_CALL_OR (C_MessageSignInit, self, CKR_SESSION_HANDLE_INVALID);
+		IN_ULONG (session)
+		IN_MECHANISM (mechanism)
+		IN_ULONG (key)
+	PROCESS_CALL;
+	END_CALL;
+}
+
+static CK_RV
+rpc_C_SignMessage (CK_X_FUNCTION_LIST *self,
+                   CK_SESSION_HANDLE session,
+                   CK_VOID_PTR parameter,
+                   CK_ULONG parameter_len,
+                   CK_BYTE_PTR data,
+                   CK_ULONG data_len,
+                   CK_BYTE_PTR signature,
+                   CK_ULONG_PTR signature_len)
+{
+	return_val_if_fail (signature_len, CKR_ARGUMENTS_BAD);
+
+	BEGIN_CALL_OR (C_SignMessage, self, CKR_SESSION_HANDLE_INVALID);
+		IN_ULONG (session)
+		IN_BYTE_ARRAY (parameter, parameter_len)
+		IN_BYTE_ARRAY (data, data_len)
+		IN_BYTE_BUFFER (signature, signature_len);
+	PROCESS_CALL;
+		OUT_BYTE_ARRAY (signature, signature_len)
+	END_CALL;
+}
+
+static CK_RV
+rpc_C_SignMessageBegin (CK_X_FUNCTION_LIST *self,
+                        CK_SESSION_HANDLE session,
+                        CK_VOID_PTR parameter,
+                        CK_ULONG parameter_len)
+{
+	BEGIN_CALL_OR (C_SignMessageBegin, self, CKR_SESSION_HANDLE_INVALID);
+		IN_ULONG (session)
+		IN_BYTE_ARRAY (parameter, parameter_len)
+
+	PROCESS_CALL;
+	END_CALL;
+}
+
+static CK_RV
+rpc_C_SignMessageNext (CK_X_FUNCTION_LIST *self,
+                       CK_SESSION_HANDLE session,
+                       CK_VOID_PTR parameter,
+                       CK_ULONG parameter_len,
+                       CK_BYTE_PTR data,
+                       CK_ULONG data_len,
+                       CK_BYTE_PTR signature,
+                       CK_ULONG_PTR signature_len)
+{
+	BEGIN_CALL_OR (C_SignMessageNext, self, CKR_SESSION_HANDLE_INVALID);
+		IN_ULONG (session)
+		IN_BYTE_ARRAY (parameter, parameter_len)
+		IN_BYTE_ARRAY (data, data_len)
+
+		assert (_msg.output);
+		/* Make sure this is in the right order */
+		assert (!_msg.signature || p11_rpc_message_verify_part (&_msg, "fy"));
+		/* Validity byte */
+		p11_rpc_buffer_add_byte (_msg.output, signature_len ? 1 : 0);
+		if (signature_len)
+			p11_rpc_buffer_add_uint32 (_msg.output, *signature_len);
+		if (p11_buffer_failed (_msg.output)) {
+			_ret = CKR_HOST_MEMORY;
+			goto _cleanup;
+		}
```

Comment:
Sorry, what I really meant was to change the protocol [signature](https://github.com/p11-glue/p11-kit/pull/463/files#diff-4c71e18b8f308838a79477ed5669813f625782daf55b2dd28714fad582afa091R245) of C_SignMessageNext from "uayayfy" to "uayayyfy" (note the extra "y" before "fy"). Then you could write this part something like:
```c
IN_BYTE (signature_present)
if (signature_present)
  IN_BYTE_BUFFER (signature, signature_len)
```